**Комментарий ревьюера**

Привет, Артём! Меня зовут Аня Титова, буду проверять твой проект по предобработке. Давай общаться на «ты». Если это неудобно, дай знать, перейдем на «вы».

У тебя уже проделана большая работа над проектом, но давай сделаем его еще лучше. Дальше ты найдешь мои комментарии — пожалуйста, не перемещай и не удаляй их. Буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b> Так выделены небольшие замечания. Одно-два таких замечания в проекте допустимы, но если их будет больше — тебе будет необходимо внести исправления. </div>

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

---

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Новые комментарии отмечу v2. Для удобства перекрашу или дополню те, что уже учтены.
</div>

---

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию" data-toc-modified-id="Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию-1">Шаг 1. Откройте файл с данными и изучите общую информацию</a></span></li><li><span><a href="#Шаг-2.-Предобработка-данных" data-toc-modified-id="Шаг-2.-Предобработка-данных-2">Шаг 2. Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-аномалий" data-toc-modified-id="Обработка-аномалий-2.1">Обработка аномалий</a></span><ul class="toc-item"><li><span><a href="#Cтолбец-children" data-toc-modified-id="Cтолбец-children-2.1.1">Cтолбец children</a></span></li><li><span><a href="#days_employed" data-toc-modified-id="days_employed-2.1.2">days_employed</a></span><ul class="toc-item"><li><span><a href="#Обработка-положительных-значений." data-toc-modified-id="Обработка-положительных-значений.-2.1.2.1">Обработка положительных значений.</a></span></li><li><span><a href="#Обработка-отрицательных-значений" data-toc-modified-id="Обработка-отрицательных-значений-2.1.2.2">Обработка отрицательных значений</a></span></li></ul></li><li><span><a href="#dob_years-—-возраст-клиента-в-годах" data-toc-modified-id="dob_years-—-возраст-клиента-в-годах-2.1.3">dob_years — возраст клиента в годах</a></span></li><li><span><a href="#education-—-уровень-образования-клиента" data-toc-modified-id="education-—-уровень-образования-клиента-2.1.4">education — уровень образования клиента</a></span></li><li><span><a href="#gender-—-пол-клиента" data-toc-modified-id="gender-—-пол-клиента-2.1.5">gender — пол клиента</a></span></li></ul></li><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.2">Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.3">Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.4">Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.5">Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6">Категоризация данных</a></span></li></ul></li><li><span><a href="#Шаг-3.-Ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Ответьте-на-вопросы-3">Шаг 3. Ответьте на вопросы</a></span></li><li><span><a href="#Шаг-4.-Общий-вывод" data-toc-modified-id="Шаг-4.-Общий-вывод-4">Шаг 4. Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5">Чек-лист готовности проекта</a></span></li></ul></div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

**Описание данных**

* __*children*__ — количество детей в семье 
* __*days_employed*__ — общий трудовой стаж в днях
* __*dob_years*__ — возраст клиента в годах
* __*education*__ — уровень образования клиента
* __*education_id*__ — идентификатор уровня образования
* __*family_status*__  — семейное положение
* __*family_status_id*__ — идентификатор семейного положения
* __*gender*__ — пол клиента
* __*income_type*__ — тип занятости
* __*debt*__ — имел ли задолженность по возврату кредитов
* __*total_income*__ — ежемесячный доход
* __*purpose*__ — цель получения кредита

In [2]:
import pandas as pd
try:
    df = pd.read_csv('/datasets/data.csv')
except:
    from pathlib2 import Path
    link = Path.cwd()/'data.csv'
    df = pd.read_csv(link)
    
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Отличное применение try-except, чтобы проект работал и локально, и на Jupyter Hubю
</div>


In [2]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
# Подсчет значений из столбцов
for columns in df.columns:
    display(df[columns][df[columns].duplicated()].value_counts(dropna=False).sort_index().to_frame())

children
-1         46
 0      14148
 1       4817
 2       2054
 3        329
 4         40
 5          8
 20        75

days_employed
NaN           2173

dob_years
0         100
19         13
20         50
21        110
22        182
23        253
24        263
25        356
26        407
27        492
28        502
29        544
30        539
31        559
32        509
33        580
34        602
35        616
36        554
37        536
38        597
39        572
40        608
41        606
42        596
43        512
44        546
45        496
46        474
47        479
48        537
49        507
50        513
51        447
52        483
53        458
54        478
55        442
56        486
57        459
58        460
59        443
60        376
61        354
62        351
63        268
64        264
65        193
66        182
67        166
68         98
69         84
70         64
71         57
72         32
73          7
74          5

education
ВЫСШЕЕ                     273
Высшее                     267
НАЧАЛЬНОЕ                   16
НЕОКОНЧЕННОЕ ВЫСШЕЕ         28
Начальное                   14
Неоконченное высшее         46
СРЕДНЕЕ                    771
Среднее                    710
высшее                    4717
начальное                  249
неоконченное высшее        667
среднее                  13749
ученая степень               3

education_id
0          5259
1         15232
2           743
3           281
4             5

family_status
Не женат / не замужем           2812
в разводе                       1194
вдовец / вдова                   959
гражданский брак                4176
женат / замужем                12379

family_status_id
0             12379
1              4176
2               959
3              1194
4              2812

gender
F   14235
M    7287

income_type
безработный                1
госслужащий             1458
компаньон               5084
пенсионер               3855
предприниматель            1
сотрудник              11118

debt
0  19783
1   1740

total_income
NaN          2173

purpose
автомобили                                  477
автомобиль                                  494
высшее образование                          452
дополнительное образование                  461
жилье                                       646
заняться высшим образованием                495
заняться образованием                       411
на покупку автомобиля                       471
на покупку подержанного автомобиля          478
на покупку своего автомобиля                504
на проведение свадьбы                       776
недвижимость                                633
образование                                 446
операции с жильем                           652
операции с коммерческой недвижимостью       650
операции с недвижимостью                    675
операции со своей недвижимостью             629
покупка жилой недвижимости                  606
покупка жилья                               646
покупка жилья для сдачи                     652
покупка жилья для семьи                     640
покупка коммерческой недвижимости           663
покупка недвижимости                        623
покупка своего жилья                        619
получение высшего образования               425
получение дополнительного образования       446
получение образования                       442
приобретение автомобиля                     461
профильное образование                      435
ремонт жилью                                611
свадьба                                     796
свой автомобиль                             479
сделка с автомобилем                        454
сделка с подержанным автомобилем            488
строительство жилой недвижимости            625
строительство недвижимости                  619
строительство собственной недвижимости      634
сыграть свадьбу                             773

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Смотри, ты применяешь to_frame(), но выводишь получившийся датафрейм c помощью print() — в этом нет особого смысла. Чтобы датафрейм отобразился красиво, используй метод display().
</div>


<div class="alert alert-info"> <b>Комментарий студента:</b> Действительно, так симпатичнее, спасибо. Убрал за ненадобностью костыльные строчки для разделения табличек.</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> 👍
</div>

In [5]:
# Подсчет уникальных значений
for column in df.columns:
    print('--------------------------------------------------------------------------')
    print('||'+ column +'||')
    print()
    print(df[column].unique())
    print()
    print('--------------------------------------------------------------------------')

--------------------------------------------------------------------------
||children||

[ 1  0  3  2 -1  4 20  5]

--------------------------------------------------------------------------
--------------------------------------------------------------------------
||days_employed||

[-8437.67302776 -4024.80375385 -5623.42261023 ... -2113.3468877
 -3112.4817052  -1984.50758853]

--------------------------------------------------------------------------
--------------------------------------------------------------------------
||dob_years||

[42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51  0 59 29 60 55 58
 71 22 73 66 69 19 72 70 74 75]

--------------------------------------------------------------------------
--------------------------------------------------------------------------
||education||

['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫ

In [6]:
401755.40047533/365 # для понимания максимальное значение дней стажа переводим в годы стажа

1100.6997273296713

**Выводы**

Названия колонок в порядке, пробелов нет, названия в snake_case, в нижнем регистре - изменений не требуется. 
В данных есть много артефактов: нуллы, пропуски, дубликаты, невалидные значения. Подробнее об аномалиях:

* __*children*__ — количество детей в семье
    1. Есть 46 значений '-1', чего быть не может.
    2. 75 значений '20' - слишком много. В России, по статистике (из инетрнета), семей из 5 и более детей - около 3%, но 20 - это рекордные значения, явный артефакт, относительно остальных значений. Имеет смысл заменить на медианное значение или удалить

***
* __*days_employed*__ — общий трудовой стаж в днях
    1. 2173 NaN
    2. Отрицательные значения
    3. Тип данных в days_employed - float64, но зачем? Количество дней удобнее представлять в int, значения после запятой, кажется, только засоряют картину, занимают больше места
    4. Максимальное значение 401755.40047533, что равно 1100.6997273296713 годам.
***    

* __*dob_years*__ — возраст клиента в годах
    1. Есть нулевые значения и их ровно 100. Не сильно много, можно будет удалить
***

* __*education*__ — уровень образования клиента
    1. Есть дубликаты с разным регистром
***

* __*gender*__ — пол клиента
    1. есть значение XNA, такого пола нет
***
* __*total_income*__ — ежемесячный доход
    1. Аналогично days_employed, 2173 NaN. Может означать отсутствие дохода а может и ошибка заполнения
    2. Возможно, неудобный тип данных float64, нам нет смысла знать доход с такой точностью
***
* __*purpose*__ — цель получения кредита
    1. Значения часто дублируются по смыслу. Здесь нужна лемматизация и удаление дубликатов


<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Почти все проблемы в данных обнаружены, прекрасно) Остается проверить полные дубли строк. 
    
Такие невозможные значения, которые многократно повторяются в столбце — это скорее всего пропуски, которые уже заполнили какими-то маркерами. Часть из них напрямую помешает исследованию. Другие не будут использоваться именно в нашем проекте, но полностью очистить и подготовить датасет — хорошая практика, ведь в работе часто приходится возвращаться к данным за дополнительными исследованиями.
</div>


## Шаг 2. Предобработка данных

### Обработка аномалий

#### Cтолбец children
Количество детей в семье
* Считаем отрицательные значения
* Убираем экстримально высокие значения '20'

In [7]:
#df[(df['children'] == -1) | (df['children']== 20)] ['children'].shape[0]
print(df[df['children'] == -1]['children'].count()) # Считаем количество '-1'
print(df[df['children'] == 20]['children'].count()) # Считаем количество '20'

47
76


Отрицательных значений: 47, а нереально больших - 76. В целом, это немного и можно убрать

In [8]:
df = df.drop(df[df['children']==-1].index)
df = df.drop(df[df['children']==20].index)

# Проверим результат удалений:
df['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Да, это пренебрежимо малая доля по сравнению с объемом нашего датасета. Но удалять мы можем полностью случайные и случайные пропуски — стоит отмечать такие моменты.
</div>


#### days_employed
Общий трудовой стаж в днях
* Убираем отрицательные значения
* Максимальное значение 401755.40047533, что равно 1100.6997273296713 года - нужно убрать. 

Но прежде чем избавляться от отрицательных значений, хочется взглянуть по отдельности, что это за значения с плюсом и минусом:

<div class="alert alert-info"> <b>Комментарий студента:</b> Обновил решение всего подблока <b>days_employed</b> по замечанию.</div>

In [9]:
print('Минимальный стаж в годах из положительных значений:', df.loc[df['days_employed']>0, 'days_employed'].min()/365)
print('Максимальный стаж в годах из положительных значений:', df.loc[df['days_employed']>0, 'days_employed'].max()/365)
print('Среднее положительных значений стажа в годах:', df.loc[df['days_employed']>0, 'days_employed'].mean()/365)
print('Медиана положительных значений стажа в годах:', df.loc[df['days_employed']>0, 'days_employed'].median()/365)
print('Число положительных значений стажа:', df.loc[df['days_employed']>0, 'days_employed'].count())

Минимальный стаж в годах из положительных значений: 900.6266317932007
Максимальный стаж в годах из положительных значений: 1100.6997273296713
Среднее положительных значений стажа в годах: 1000.0709627430559
Медиана положительных значений стажа в годах: 1000.7852675329941
Число положительных значений стажа: 3431


Все положительные данные выглядит слишком большими, минимум: 900 лет, максимум: 1100, медиана и среднее отличаются несущественно. Данные явно  ошибочны.

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Верное замечание</div>

Посмотрим аналогично на отрицательные значения.

In [10]:
print('Минимальный стаж в годах из отрицательных значений:', df.loc[df['days_employed']<0, 'days_employed'].abs().min()/365)
print('Максимальный стаж в годах из отрицательных значений:', df.loc[df['days_employed']<0, 'days_employed'].abs().max()/365)
print('Среднее отрицательных значений стажа в годах:', df.loc[df['days_employed']<0, 'days_employed'].abs().mean()/365)
print('Медиана отрицательных значений стажа в годах:', df.loc[df['days_employed']<0, 'days_employed'].abs().median()/365)
print('Число отрицательных значений стажа:', df.loc[df['days_employed']<0, 'days_employed'].count())

Минимальный стаж в годах из отрицательных значений: 0.06614146093282515
Максимальный стаж в годах из отрицательных значений: 50.38068465909146
Среднее отрицательных значений стажа в годах: 6.44836710137888
Медиана отрицательных значений стажа в годах: 4.46696874194262
Число отрицательных значений стажа: 15809


А вот отрицательные значения больше похожи на правду, нет аномальных значений. Их менять не следует, кроме как убрать знак минус

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> 👍
</div>

##### Обработка положительных значений. 
Весьма вероятно, они стали в 100 раз больше после импорта из какой-то системы или при конвертации. Поэтому, сначала уменьшим положителньые значения в 100 раз, а потом уже уберим знак минус у отрицательных

In [11]:
df.loc[df['days_employed'] > 0, 'days_employed'] = df['days_employed']/100
# Проверка:
print('Минимальный стаж в годах из положительных значений:', df.loc[df['days_employed']>0, 'days_employed'].min()/365)
print('Максимальный стаж в годах из положительных значений:', df.loc[df['days_employed']>0, 'days_employed'].max()/365)
print('Среднее положительных значений стажа в годах:', df.loc[df['days_employed']>0, 'days_employed'].mean()/365)
print('Медиана положительных значений стажа в годах:', df.loc[df['days_employed']>0, 'days_employed'].median()/365)
print('Число положительных значений стажа:', df.loc[df['days_employed']>0, 'days_employed'].count())

Минимальный стаж в годах из положительных значений: 9.006266317932008
Максимальный стаж в годах из положительных значений: 11.006997273296712
Среднее положительных значений стажа в годах: 10.000709627430561
Медиана положительных значений стажа в годах: 10.007852675329941
Число положительных значений стажа: 3431


<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Ок, это возможный способ восстановить значения.
</div>

##### Обработка отрицательных значений
Теперь просто уберем отрицательные значения методом abs()

In [12]:
print('Количество отрицательных значений до обработки', df[df['days_employed'] < 0].shape[0])
df['days_employed'] = df['days_employed'].abs() # Убираем отрицательные значения
print('Количество отрицательных значений ПОСЛЕ обработки', df[df['days_employed'] < 0].shape[0]) #Проверяем наличие отрицательных значений

Количество отрицательных значений до обработки 15809
Количество отрицательных значений ПОСЛЕ обработки 0


In [13]:
df.sort_values(by = 'days_employed', ascending = False).head(10)

children  days_employed  dob_years education  education_id  \
16335         1   18388.949901         61   среднее             1   
4299          0   17615.563266         61   среднее             1   
7329          0   16593.472817         60    высшее             0   
17838         0   16264.699501         59   среднее             1   
16825         0   16119.687737         64   среднее             1   
3974          0   15835.725775         64   среднее             1   
1539          0   15785.678893         59    высшее             0   
4321          0   15773.061335         61   среднее             1   
7731          0   15618.063786         64   среднее             1   
15675         0   15410.040779         65    высшее             0   

               family_status  family_status_id gender income_type  debt  \
16335        женат / замужем                 0      F   сотрудник     0   
4299         женат / замужем                 0      F   компаньон     0   
7329         женат / замужем                 0      F   сотрудник     0   
17838        женат / замужем                 0      F   сотрудник     0   
16825        женат / замужем                 0      F   сотрудник     0   
3974        гражданский брак                 1      F   компаньон     0   
1539   Не женат / не замужем                 4      F   сотрудник     0   
4321        гражданский брак                 1      F   сотрудник     0   
7731         женат / замужем                 0      F   компаньон     0   
15675        женат / замужем                 0      F   сотрудник     0   

        total_income                                purpose  
16335  186178.934089               операции с недвижимостью  
4299   122560.741753                          покупка жилья  
7329   124697.846781           заняться высшим образованием  
17838   51238.967133                  на покупку автомобиля  
16825   91527.685995             покупка жилой недвижимости  
3974    96858.531436                        сыграть свадьбу  
1539   119563.851852  операции с коммерческой недвижимостью  
4321   205868.585780                                свадьба  
7731   296525.358574                     высшее образование  
15675  188800.068859             покупка жилой недвижимости

In [14]:
df.sort_values(by = 'days_employed', ascending = False).tail(10)

children  days_employed  dob_years            education  education_id  \
21415         0            NaN         54              среднее             1   
21423         0            NaN         63              среднее             1   
21426         0            NaN         49              среднее             1   
21432         1            NaN         38  неоконченное высшее             2   
21463         1            NaN         35               высшее             0   
21489         2            NaN         47              Среднее             1   
21495         1            NaN         50              среднее             1   
21497         0            NaN         48               ВЫСШЕЕ             0   
21502         1            NaN         42              среднее             1   
21510         2            NaN         28              среднее             1   

               family_status  family_status_id gender income_type  debt  \
21415        женат / замужем                 0      F   пенсионер     0   
21423        женат / замужем                 0      M   пенсионер     0   
21426        женат / замужем                 0      F   сотрудник     1   
21432  Не женат / не замужем                 4      F   сотрудник     0   
21463       гражданский брак                 1      M   сотрудник     0   
21489        женат / замужем                 0      M   компаньон     0   
21495       гражданский брак                 1      F   сотрудник     0   
21497        женат / замужем                 0      F   компаньон     0   
21502        женат / замужем                 0      F   сотрудник     0   
21510        женат / замужем                 0      F   сотрудник     0   

       total_income                           purpose  
21415           NaN                 операции с жильем  
21423           NaN              сделка с автомобилем  
21426           NaN                      недвижимость  
21432           NaN                 операции с жильем  
21463           NaN             на проведение свадьбы  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Постарайся не вызывать весь датафрейм целиком — так можно повесить Jupyter Hub. Лучше используй для вывода первых и последних строк методы head() и tail().

<b>Комментарий ревьюера v2:</b> Сделано</div>


<div class="alert alert-info"> <b>Комментарий студента:</b> Ок, сделал 2 вызова датафрейма с head() и tail()</div>

In [15]:
print('Средний стаж в годах:', df['days_employed'].mean()/365) #получаем среднее значение стажа в годах
print('Медиана стажа в годах:', df['days_employed'].median()/365) #получаем медианное значение стажа

Средний стаж в годах: 7.081843567433105
Медиана стажа в годах: 6.014387921983277


__*Старое решение, выше я немного поменял и теперь оно неактуально*__ 

~Высокие значения стажа лишены логики, нужно ограничить неким максимумом. Сведения о максимальных значения стажа найти не~ ~удалось. Из открытых источников средний стаж жителя России до выхода на пенсию равен 34 годам. Но, нас интересуют значения,~ ~приближенные к максимальным. Из открытых же истончиков берем примерную цифру средней продолжительности жизни в России на~ ~2020год: __72 года__. Официально можно работать с __16 лет__.~

~Соответственно, ограничим стаж числом 72 - 16 = __56__ ~
~Но для начала, надо понять, сколько записей будет отброшено:~

__*Новое решение*__
Немного поменял решение: обработал все положительные значения, которые были запмсаны с систематической ошибкой - увеличил их в 100 раз. И только потом убрал знак минус у отрциательных, значения которых в кажутся адекватными. Получилось почти то же самое, судя по медиане и средней стажа.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Здесь странно решение скорректировать только часть значений. Подумай, откуда могли взяться эти аномалии. А если «нормальные» дни стажа тоже выгружены с ошибкой и только кажутся корректными? Сами по себе данные так не бьются — явно произошла ошибка при конвертации. То, что с этим столбцом точно делали — переводили из даты трудоустройства в дни стажа. Скорее всего, проблема касается всех значений, которые не были отрицательными.

<b>Комментарий ревьюера v2:</b> Есть правки</div>


<div class="alert alert-info"> <b>Комментарий студента:</b> Я изначально предположил, что ошибка из-за форматирования (переноса данных из одной системы в другую, например). Но если взять все изначально неотрицательные значения, то их и было 3431, как раз мое первое решение их и обработало. Група неотрицательных значений вся сильно больше, минимум 900 лет, максимум 1100; Тогда как группа изначально отрицательных значений за исключением знака была больше похожа на правду, там максиум - 50 лет, минимум 24 дня. 

Немного поменял решение, но там ведь тоже меняется только часть данных: обработал все положительные значения, только потом убрал знак минус у отрциательных. Получилось то же самое. Но изначально мне это в голову не пришло, тк я сразу избавился от отрицательных значений и уже работал со всеми положительными.
Спасибо за несколько иной подход, я не подумал отдельно посмотреть на отрицательные значения и отдельно на положителньые (и что именно они могли быть аномально большими) Забавно, что в моем случае, результат все равно один: поправилось то же число стажа и средний стаж 1ым решенеим и 2ым решением был равен около 7.08, но это, кажется, мне повезло.
Старое решение убирать не стал, просто закомментировал код и оставил комменты
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Отличная работа. Да, получается, что обе группы аномалий были так распределены, что вариант с рассчитанной нормой стажа тоже дал верный результат. Важно было убедиться, что мы не оставили ничего без обработки.
</div>

#### dob_years — возраст клиента в годах
Заменим нулевые значения на NaN для дальнейшей их обработки

In [16]:
import numpy as np
df.loc[df['dob_years'] == 0, 'dob_years'] = np.nan
df[df['dob_years'] == 0] #Проверяем наличие нулевых значений

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

#### education — уровень образования клиента
Приведем к нижнему регистру

In [17]:
df['education'].value_counts() # Еще раз, глянем как сейчас обстоят дела:

среднее                13667
высшее                  4698
СРЕДНЕЕ                  766
Среднее                  703
неоконченное высшее      665
ВЫСШЕЕ                   271
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

In [18]:
df['education'] = df['education'].str.lower() # Приводим к нижнему регистру все значения
df['education'].value_counts()

среднее                15136
высшее                  5237
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> 👍
</div>


#### gender — пол клиента
Убрать значение XNA, такого пола нет

Сначала еще раз посчитаем, сколько значений XNA сейчас:

In [19]:
df.loc[df['gender'] == 'XNA']['gender'].count() # считаем, сколько значений XNA сейчас

1

In [20]:
df = df.drop(df[df['gender']=='XNA'].index)
df.loc[df['gender'] == 'XNA']['gender'].count() # проверяем количество XNA после обработки

0

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Ок
</div>


### Обработка пропусков

In [21]:
print(df.isnull().sum()) # Считаем количество пропусков в датафрейме

children               0
days_employed       2162
dob_years            100
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2162
purpose                0
dtype: int64


In [22]:
df.isna().sum()

children               0
days_employed       2162
dob_years            100
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2162
purpose                0
dtype: int64

Много пропусков, но только в 2 столбцах: трудовой стаж (days_employed) и ежемесячный доход (total_income), так же немного - возраст клиента (dob_years), но это нулевые значения. 

Пропуски в dob_years скорее всего ошибочно незаполненные данные. В случае с трудовым стажем и доходом пропуски могут быть случайными и нет.  Одна из причин - клиент никогда не работал и сейчас безработный. Либо же, работа с черной зп. Либо, банальное отсутсвие желания сообщать информацию о своих доходах. В любом случае, количество пропусков велико, что не позволяет нам их просто удалить. Что делать?
* Как вариант, можно избавиться от всех данных с пропусками. Но пропусков много, мы сильно потеряем в количестве данных
* Можно заменить данные на среднее арифметическое или медиану. Но заполнять одним значением 10% данных - грозит ошибкой
* Можно попытаться заменить на среднее, сгруппировав по другим характеристикам

Попробуем 3ий вариант:

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Да, вариант с заменой по группам — оптимальный.
</div>


In [23]:
print('Пропуски в датафрейме:' '\n')
print(df.isna().sum())
df['days_employed'] = df.groupby(['income_type', 'children', 'education'])['days_employed'].transform(lambda x: x.fillna(x.mean()))
df['dob_years'] = df.groupby(['income_type', 'children', 'education'])['dob_years'].transform(lambda x: x.fillna(x.mean()))
df['total_income'] = df.groupby(['income_type','education'])['total_income'].transform(lambda x: x.fillna(x.mean()))
print('\n' 'Пропуски в датафрейме после обработки:')
df.isna().sum()

Пропуски в датафрейме:

children               0
days_employed       2162
dob_years            100
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2162
purpose                0
dtype: int64

Пропуски в датафрейме после обработки:


children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

**Вывод**

От пропусков избавились 3им способом. В итоге:
* dob_years - средним значением по группам тип занятости, наличие детей и образование, тк по ним можно сделать вывод о возрасте;
* days_employed - средним значением по группам: тип занятости, наличие детей и образование;
* total_income - средним значением по группам: тип занятости, образование (если взять еще и по возрасту, то у меня остаются пропуски)

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Пропуски обработаны. Ты сразу предполагаешь подходящие группы для замены, а еще можно поискать лучшие варианты экспериментально. Например, делаем разные группировки и считаем средний доход в этих группах. Если среднее получается похожим от группы к группе, то такие категории слабо связаны с доходом. А где больше всего разнится — там категории, для которых лучше всего и брать характерные значения.
    
</div>


### Замена типа данных

Как сейчас обстоят дела?

In [24]:
df.dtypes 

children              int64
days_employed       float64
dob_years           float64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21401 entries, 0 to 21524
Data columns (total 12 columns):
children            21401 non-null int64
days_employed       21401 non-null float64
dob_years           21401 non-null float64
education           21401 non-null object
education_id        21401 non-null int64
family_status       21401 non-null object
family_status_id    21401 non-null int64
gender              21401 non-null object
income_type         21401 non-null object
debt                21401 non-null int64
total_income        21401 non-null float64
purpose             21401 non-null object
dtypes: float64(3), int64(4), object(5)
memory usage: 2.1+ MB


In [26]:
df['days_employed'] = df['days_employed'].astype('int32')
df['total_income'] = df['total_income'].astype('int32')
df['debt'] = df['debt'].astype('bool')
df['dob_years'] = df['dob_years'].astype('int64')

#Необязательные изменения:
#df['children'] = df['children'].astype('int8')
#df['family_status_id'] = df['family_status_id'].astype('int8') # Можно заменить на Int(8), тк там значения от 1-4. int(64) - это оверехед по памяти
#df['education_id'] = df['education_id'].astype('int8')
print('Типы данных после изменения:' '\n','\n', df.dtypes)

Типы данных после изменения:
 
 children             int64
days_employed        int32
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                  bool
total_income         int32
purpose             object
dtype: object


**Вывод**
* Изменил тип __days_employed__ с float(64) на int(32), так как мы имеем дело с днями, цифры после запятой излишни, неудобны для восприятия и могут приводить к неожиданным последствиям при манипуляциях с данными
* Абсолютно аналогично изменил тип с float(64) на int(32) __total_income__ 
* Для __debt__ смена типа на bool была обоснована тем, что там всего 2 уникальных значения: 1 или 0
* __dob_years__ изначально был в int(64)? но при заполнении средними значениями стал float(64). Вернул тип int(64)

Было желание ради оптимизации использования памяти изменить типы с int(64) на int(8) или int(16) в столбцах __children, family_status_id, education_id__ тк значения с небольшим диапазоном. Но, думаю, заморачиваться в данном случае не стоит, у нас нет сложных алгоритмов, работа идет не с большими данными, так что это излишне.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Типы привели, отлично. Позаботиться о потреблении памяти можно и ради тренировки) Но самое главное — сделать целочисленными переменные, в которых такой тип ожидается.
</div>


### Обработка дубликатов

In [27]:
print('Дубли в датафрейме:' '\n')
print(df.duplicated().sum())
df = df.drop_duplicates().reset_index(drop = True)
print('\n''Дубли в датафрейме после удаления дубликатов:' '\n')
df.duplicated().sum()

Дубли в датафрейме:

71

Дубли в датафрейме после удаления дубликатов:



0

**Вывод**
В 2.1.4 были удалены дубли (приведением к одному регистру), вызванные написанием одних и тех же значений в столбце __education__ разным регистром

Плюс, был найден 71 оставшийся дубликат, который был удален обычным методом _astype_ 

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Отлично, дубли устранили. Можешь ради интереса посмотреть, сколько было полных дублей строк до приведения регистра в образовании.
</div>


### Лемматизация

In [28]:
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Как мы заметили еще в Шаге1, здесь нужна лематизация, чтоб убрать дубликаты. Из выделив все уникальные значения, явно проглядываются 4 цели:
* Организация свадьбы
* Покупка автомобиля
* Операция с недвижимостью
* Образование
Нужно выделить леммы и привести столбец purpose к новым значениям

In [29]:
# Старое, немного поправленное рабочее решение
from pymystem3 import Mystem
from collections import Counter
m = Mystem()

def lem_func(row):
    lemmas = m.lemmatize(row['purpose'])
    return lemmas

purpose_row = df.apply(lem_func, axis=1).reset_index(drop=True)
purpose_row.explode().duplicated()

for counter_1 in range(len(df)): 
    lemm = purpose_row[counter_1]
    for counter_2 in range(len(lemm)):
        if lemm[counter_2] == 'свадьба':
            df.loc[counter_1,'purpose'] = 'организация свадьбы'
        if lemm[counter_2] == 'автомобиль':
            df.loc[counter_1,'purpose'] = 'покупка автомобиля'
        if lemm[counter_2] == 'жилье':
            df.loc[counter_1,'purpose'] = 'операция с недвижимостью'
        if lemm[counter_2] == 'недвижимость':
            df.loc[counter_1,'purpose'] = 'операция с недвижимостью'
        if lemm[counter_2] == 'образование':
            df.loc[counter_1,'purpose'] = 'образование'
            
print(df['purpose'].unique())

['операция с недвижимостью' 'покупка автомобиля' 'образование'
 'организация свадьбы']


In [30]:
""" # Новое решение, которое не смогло :(
from pymystem3 import Mystem
from collections import Counter
m = Mystem()

def lem_func(row):
    lemmas = m.lemmatize(row['purpose'])
    return lemmas

purpose_row = df.apply(lem_func, axis=1).reset_index(drop=True)
print(purpose_row.explode())
#print(purpose_row)

def purpose_change(row): # Фукнция для подстановки лемм в столбец
    lemmas_row = m.lemmatize(purpose)
    if lemmas_row == 'свадьба':
        return 'организация свадьбы'
    if lemmas_row == 'автомобиль':
        return 'покупка автомобиля'
    if lemmas_row == 'жилье':
        return 'операция с недвижимостью'
    if lemmas_row == 'недвижимость':
        return 'операция с недвижимостью'
    if lemmas_row == 'образование':
        return 'образование'
row = pd.Series(data=row_values, index=row_columns)
df['purpose']= df.apply(purpose_change)        
print(df.head(20))
"""

" # Новое решение, которое не смогло :(\nfrom pymystem3 import Mystem\nfrom collections import Counter\nm = Mystem()\n\ndef lem_func(row):\n    lemmas = m.lemmatize(row['purpose'])\n    return lemmas\n\npurpose_row = df.apply(lem_func, axis=1).reset_index(drop=True)\nprint(purpose_row.explode())\n#print(purpose_row)\n\ndef purpose_change(row): # Фукнция для подстановки лемм в столбец\n    lemmas_row = m.lemmatize(purpose)\n    if lemmas_row == 'свадьба':\n        return 'организация свадьбы'\n    if lemmas_row == 'автомобиль':\n        return 'покупка автомобиля'\n    if lemmas_row == 'жилье':\n        return 'операция с недвижимостью'\n    if lemmas_row == 'недвижимость':\n        return 'операция с недвижимостью'\n    if lemmas_row == 'образование':\n        return 'образование'\nrow = pd.Series(data=row_values, index=row_columns)\ndf['purpose']= df.apply(purpose_change)        \nprint(df.head(20))\n"

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b> Не стоит перебирать весь датафрейм построчно циклом for — это медленно. Лучше для такой операции тоже написать функцию и передать ее в apply().
</div>


<div class="alert alert-info"> <b>Комментарий студента:</b> Да сразу было понятно, что решение неизящное, но к сожалению, пока не удалось сделать за вменяемое время 2ую функцию и чтоб все взлетело :(</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Смотри, метод m.lemmatize можно сразу передать в apply() и так получить новый столбец с леммами. А затем пишем функцию, которая на условных операторах проверяет вхождение ключевого слова в столбец с леммами и возвращает категорию — так получим столбец с самими категориями.
</div>

In [31]:
df['purpose'].tail(15)

21315          покупка автомобиля
21316    операция с недвижимостью
21317                 образование
21318    операция с недвижимостью
21319    операция с недвижимостью
21320                 образование
21321    операция с недвижимостью
21322          покупка автомобиля
21323          покупка автомобиля
21324    операция с недвижимостью
21325    операция с недвижимостью
21326          покупка автомобиля
21327    операция с недвижимостью
21328          покупка автомобиля
21329          покупка автомобиля
Name: purpose, dtype: object

**Вывод**

Из всего столбца __purpose__ выделены 4 основные категории, которые теперь являются целями кредита, дублей по значению больше нет

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Отлично, лемматизация выполнена.
</div>


### Категоризация данных

Хотелось бы выделить категории по:
* __доходу__, поэтому разделим людей на разные доходы: низкий, средний, высокий, очень высокий 
* __количеству детей__ - есть до 2 детей, многодетная семья, нет
* __семейному положению__ - и __цели кредита__ немногочисленны, по этому, сами по себе категории

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Да, хороший подход.
</div>


In [32]:
print(df['total_income'].median())
print(df['total_income'].mean())

df.groupby(['income_type', 'education'])['total_income'].mean()

152411.0
167509.12930145336


income_type      education          
безработный      высшее                 202722.000000
                 среднее                 59956.000000
в декрете        среднее                 53829.000000
госслужащий      высшее                 197102.758684
                 начальное              184056.000000
                 неоконченное высшее    172476.365385
                 среднее                153874.790036
                 ученая степень         111392.000000
компаньон        высшее                 242575.235228
                 начальное              165056.566667
                 неоконченное высшее    197873.071685
                 среднее                179573.143919
пенсионер        высшее                 170667.297909
                 начальное              111314.424242
                 неоконченное высшее    138311.682927
                 среднее                131772.754199
                 ученая степень         177088.500000
предприниматель  высшее                 49916

// Что-то многовато в среднем получают у вас люди, даже для Москвы :)

Для начала, категоризируем доход в категории "низкий", "средний", "высокий" и "очень высокий" и количество детей в "есть дети", "бездетная семья" и "многодетная семья"

In [33]:
def income_func(income):
    if income <= 50000:
        return 'низкий'
    if income <= 160000:
        return 'средний'
    if income <= 300000:
        return 'высокий'
    return 'очень высокий'
df['income_level'] = df['total_income'].apply(income_func)

def child_func(childs):
    if childs == 0:
        return 'бездетная семья'
    if childs <=2:
        return 'есть дети'
    if childs > 2:
        return 'многодетная семья'
df['child_level'] = df['children'].apply(child_func)

Так как у нас неплательщик это __1__, а плательщик __0__ - можно просто просуммировать значения. __sum__ вместо __count__. А если посчитать __mean__ - получится как раз доля неплательщиков.
Сгруппируем долю неплательщиков по: 
* уровню дохода
* наличию детей
* целям кредита
* семейному положению


In [34]:
print('Доля неплательщиков по уровню доходу:', '\n', df.groupby('income_level')['debt'].mean())
print('\n','Доля неплательщиков по наличию детей:','\n', df.groupby('child_level')['debt'].mean())
print('\n','Доля неплательщиков по целям кредита:','\n', df.groupby('purpose')['debt'].mean())
print('\n','Доля неплательщиков по семейному положению:','\n', df.groupby('family_status')['debt'].mean())

Доля неплательщиков по уровню доходу: 
 income_level
высокий          0.076295
низкий           0.061995
очень высокий    0.071913
средний          0.086079
Name: debt, dtype: float64

 Доля неплательщиков по наличию детей: 
 child_level
бездетная семья      0.075444
есть дети            0.093003
многодетная семья    0.081579
Name: debt, dtype: float64

 Доля неплательщиков по целям кредита: 
 purpose
образование                 0.092528
операция с недвижимостью    0.072558
организация свадьбы         0.079118
покупка автомобиля          0.093480
Name: debt, dtype: float64

 Доля неплательщиков по семейному положению: 
 family_status
Не женат / не замужем    0.097639
в разводе                0.070648
вдовец / вдова           0.066246
гражданский брак         0.093153
женат / замужем          0.075606
Name: debt, dtype: float64


<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Отлично, категории выделили, и метрику тоже выбрали правильную. 
</div>


**Вывод**

Выделил категории дохода: 
* доход - "низкий", "средний", "высокий" и "очень высокий"
* количество детей - в "есть дети", "бездетная семья" и "многодетная семья"

Заодно, подсчитал долю неплательщиков в каждой категории

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Постарайся подводить итоги в подпункте с выводом, они как раз для этого в шаблоне :)
</div>


<div class="alert alert-info"> <b>Комментарий студента:</b> Добавил вывод</div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [35]:
df.pivot_table(index='child_level', values='debt', aggfunc='mean')

debt
child_level                
бездетная семья    0.075444
есть дети          0.093003
многодетная семья  0.081579

**Вывод**

Доля неплательщиков без детей намного ниже, чем у тех, у кого есть дети (до 2 детей), и ниже, чем у многодетных семей (более 2). 

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Верно.
</div>


- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [36]:
df.pivot_table(index='family_status', values='debt', aggfunc='mean')

debt
family_status                  
Не женат / не замужем  0.097639
в разводе              0.070648
вдовец / вдова         0.066246
гражданский брак       0.093153
женат / замужем        0.075606

**Вывод**

Доля неплательщиков самая высокая у неженатых/незамужних (0.097) и тех, кто в гражданском браке (0.093). Самая низкая - у вдовцов/вдов (0.066) и чуть выше тех, кто в разводе (0.07) и женат/замужем (0.076)

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Здесь забыл переписать вывод, хотя в итоговом выводе он есть)

<b>Комментарий ревьюера v2:</b> 👍</div>


<div class="alert alert-info"> <b>Комментарий студента:</b> Да, шальной копипаст и невнимательность, исправил</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [37]:
df.pivot_table(index='income_level', values='debt', aggfunc='mean')

debt
income_level           
высокий        0.076295
низкий         0.061995
очень высокий  0.071913
средний        0.086079

**Вывод**

Немного странная зависимость. Наибольшая доля неплательщиков - со средним доходом - 0.086. Наименьшая - у клиентов с низким доходом - 0.06. У клиентов с высоким и очень высоким доходом - показатели средние

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Здесь метрика очень чувствительна к способу категоризации по доходу. Можешь посмотреть, насколько равномерно поделились клиенты по группам. Если какая-то из них крайне малочисленна, то ее нельзя учитывать в анализе. Но в целом примерно так и получается, что в нашем датасете у среднего уровня дохода наибольшая доля просрочек.
</div>


- Как разные цели кредита влияют на его возврат в срок?

In [38]:
df.pivot_table(index='purpose', values='debt', aggfunc='mean')

debt
purpose                           
образование               0.092528
операция с недвижимостью  0.072558
организация свадьбы       0.079118
покупка автомобиля        0.093480

**Вывод**

Наибольшая доля неплательщиков - среди тех, кто берет кредит на образование и автомобиль. Меньшая - у тех, кто берет на свадьбу и на вопросы, связанные с недвижимостью

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> 👍
</div>


<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Здорово, что используешь pivot_table() для вывода финальных результатов — это полезный метод, и его важно освоить.
</div>


## Шаг 4. Общий вывод

Резюмирая, мы попытались увидеть зависимость возврата кредита от __дохода, цели кредита, семейного положения и количества детей__.

* __*Количество детей*__
    Наименьшая доля неплательщиков (0.075) - у бездетных
    Наибольшая доля неплательщиков (0.09) - у тех, у кого есть 1 и 2 ребенка
    У многодетных (более 2 детей) показатель средний (0.08)

* __*Семейное положение*__
    Наименьшая доля неплательщиков (0.066) - у тех, кто потерял супруга
    Наибольшая доля неплательщиков (0.097) - у тех, кто не женат/не замужем и в гражданском браке (0.093)

* __*Уровень дохода*__ есть неочевидная зависимость
    Те, у кого самый низкий доход реже всего являются неплательщиками (0.06), тогда как люди со средним доходом чаще всего становятся должниками (0.086)

* __*Цель кредита*__ довольно полярно расположились причины кредита:  
    Наименьшая доля неплательщиков: (0.07) - у тех, кто берет кредит на недвижимость. Никто не хочет остаться без жилья. 
    Наибольшая доля неплательщиков: (0.093) - у тех, кто берет кредит на автомобиль и образование (0.092). Возможно, автомобиль часто берется не по средствам, а студенты не всегда могут погасить кредит на образование.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Отлично, на вопросы исследования ответили. Здесь можно округлить метрики до 2 знаков после запятой — большая точность нам не слишком нужна для качественного вывода, а восприятию мешает.
</div>


<div class="alert alert-info"> <b>Комментарий студента:</b> Округлил до 2-3 знаков после запятой </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Хорошо, теперь легче прочитать цифры.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию" data-toc-modified-id="Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Откройте файл с данными и изучите общую информацию</a></span></li><li><span><a href="#Шаг-2.-Предобработка-данных" data-toc-modified-id="Шаг-2.-Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2. Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-аномалий" data-toc-modified-id="Обработка-аномалий-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка аномалий</a></span><ul class="toc-item"><li><span><a href="#Cтолбец-children" data-toc-modified-id="Cтолбец-children-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Cтолбец children</a></span></li><li><span><a href="#days_employed" data-toc-modified-id="days_employed-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>days_employed</a></span></li><li><span><a href="#dob_years-—-возраст-клиента-в-годах" data-toc-modified-id="dob_years-—-возраст-клиента-в-годах-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>dob_years — возраст клиента в годах</a></span></li><li><span><a href="#education-—-уровень-образования-клиента" data-toc-modified-id="education-—-уровень-образования-клиента-2.1.4"><span class="toc-item-num">2.1.4&nbsp;&nbsp;</span>education — уровень образования клиента</a></span></li><li><span><a href="#gender-—-пол-клиента" data-toc-modified-id="gender-—-пол-клиента-2.1.5"><span class="toc-item-num">2.1.5&nbsp;&nbsp;</span>gender — пол клиента</a></span></li></ul></li><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Шаг-3.-Ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Ответьте-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 3. Ответьте на вопросы</a></span></li><li><span><a href="#Шаг-4.-Общий-вывод" data-toc-modified-id="Шаг-4.-Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Шаг 4. Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

<div class="alert alert-success">
<b>Итоговый комментарий ревьюера:</b> Отличная работа. Аккуратно изучаешь и подготавливаешь данные. Описываешь, как принимались решения по обработке и интерпретируешь результаты исследования. 
    
Критичных замечаний нет, но пока многовато мелких. Само по себе каждое из них может быть несущественно, но в массе они снижают уровень работы. Пожалуйста, внеси правки хотя бы в часть из них и присылай проект на дополнительное ревью.
</div>


<div class="alert alert-success">
<b>Итоговый комментарий ревьюера v2:</b> Почти все замечания учтены, проект принимаю) По функциям вместо построчного перебора постаралась подсказать вариант. Ты можешь посмотреть еще примеры в тренажере — тема Категоризация данных, уроки Классификация по возрастным группам, Функция для одной строки. Здесь у нас небольшой датафрейм, но в целом навык очень пригодится.
</div>
